core/src/main/java/org/apache/iceberg/actions/RewriteDeleteStrategy.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.actions;
+
+import java.util.Map;
+import java.util.Set;
+import org.apache.iceberg.DeleteFile;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.Table;
+
+public interface RewriteDeleteStrategy {
+
+  /**
+   * Returns the name of this rewrite deletes strategy
+   */
+  String name();
+
+  /**
+   * Returns the table being modified by this rewrite strategy
+   */
+  Table table();
+
+  /**
+   * Select the deletes to rewrite.
+   *
+   * @param dataFiles iterable of FileScanTasks for data files in a given partition
+   * @return iterable of original delete file to be replaced.
+   */
+  Iterable<DeleteFile> selectDeletesToRewrite(Iterable<FileScanTask> dataFiles);
```

Comment:
Why accept `FileScanTask` here? Will the delete action only find delete files based on a scan?

Reply:
This was discussed in https://github.com/apache/iceberg/pull/2841#discussion_r672569391. It is in order to align with `RewriteStrategy`.

Reply:
I think there is a third use case two: simply compacting position deletes.

Reply:
Also, will position -> data be something we take here or in `RewriteDataFiles`? I can see it being part of the data compaction. Like if we detect that certain data files have too many matching delete files, we want to rewrite them.

Reply:
I agree. This is the major compaction that we mentioned before.  And I think this may be already available through `RewriteDataFiles`.

Reply:
@jackye1995,  In order to make API more generic, maybe we can go back to use empty args and let the strategy implementation maintain the map like the implementation [here]( https://github.com/apache/iceberg/pull/2364/files#diff-b31571d37b1152b422d32af62978f2954359254ac11a7f4338e3116b1b262c4cR108). So that users could choose the deletes in their own way, for example, read from manifest directly.  And since we are rewriting deletes, it looks more natural to return `Iterable<DeleteFile>`.

@rdblue @jackye1995 WDYT?

Reply:
@aokolnychyi thanks for the feedback!

> I think there is a third use case: simply compacting position deletes.

Yes agree, that maps to the Hive minor compaction as well.

> will position -> data be something we take here or in RewriteDataFiles? I can see it being part of the data compaction. Like if we detect that certain data files have too many matching delete files, we want to rewrite them.

I remember I had a discussion around this with Russell when he was implementing bin packing. When we rewrite data files, yes position delete information is merged to that, but delete files are not dropped, because the delete file might still reference other files that are not included during bin-packing. We discussed the possibility for also adding that feature in the data compaction action, the conclusion was to have another action specifically for this use case to (1) not make data compaction too complicated, (2) have separated actions for different purposes, (3) have better SLA control for different actions.